src/hooks/useValidateRepository.ts

```diff
@@ -0,0 +1,188 @@
+            uri: 'error',
+            token: getFieldValue('token') ? 'error' : '',
+            username: getFieldValue('username') ? 'error' : '',
+          });
+          return;
+        }
+        if (errorDetail.includes(uriSearchString)) {
+          setValidationState({
+            message: getErrorMessage(errorDetail, uriSearchString),
+            uri: 'error',
+          });
+          return;
+        }
+
+        if (errorDetail.includes(branchSearchString)) {
+          setValidationState({
+            message: getErrorMessage(errorDetail, branchSearchString),
+            uri: 'success',
+            token: getFieldValue('token') ? 'success' : '',
+            username: getFieldValue('username') ? 'success' : '',
+            branch: 'error',
+          });
+          return;
+        }
+
+        if (errorDetail.includes(pathSearchString)) {
+          setValidationState({
+            message: getErrorMessage(errorDetail, pathSearchString),
+            uri: 'success',
+            token: getFieldValue('token') ? 'success' : '',
+            username: getFieldValue('username') ? 'success' : '',
+            branch: getFieldValue('branch') ? 'success' : '',
+            path: 'error',
+          });
+          return;
+        }
+
+        setValidationState({
+          message: 'Unknown error',
+          uri: 'error',
+        });
+      } else {
+        setValidationState({
+          uri: 'success',
+          token: getFieldValue('token') ? 'success' : '',
+          username: getFieldValue('username') ? 'success' : '',
+          branch: getFieldValue('branch') ? 'success' : '',
+          path: getFieldValue('path') ? 'success' : '',
+        });
+      }
+    },
+    [setValidationState, getFieldValue]
+  );
+
+  const getValidationPayload = useCallback(() => {
+    return {
+      type: 'git',
+      ...fieldsNames.reduce((acc, curr) => {
+        if (curr === 'token' && getFieldValue(curr) === dummySecret && getFieldValue('tokenSecret')) {
+          return {
+            ...acc,
+            tokenSecret: getFieldValue('tokenSecret'),
+          };
+        }
+
+        if (curr === 'username' && getFieldValue(curr) === dummySecret && getFieldValue('usernameSecret')) {
+          return {
+            ...acc,
+            usernameSecret: getFieldValue('usernameSecret'),
+          };
+        }
+        return {...acc, [curr]: getFieldValue(curr)};
+      }, {}),
+    };
+  }, [getFieldValue]);
+
+  useDebounce(
+    () => {
+      if (!getFieldValue('uri')) {
+        return;
+      }
+
+      setValidationState({
+        uri: 'loading',
+        token: getFieldValue('token') ? 'loading' : '',
+        username: getFieldValue('username') ? 'loading' : '',
+      });
+
+      validateRepository(getValidationPayload()).then((res: any) => {
+        handleResponse(res);
+      });
+    },
+    300,
+    [getFieldValue('uri'), getFieldValue('token'), getFieldValue('username')]
+  );
+
+  useDebounce(
```

Comment:
Shouldn't all these 3 debounces be combined together? These are done for exactly same thing, and each of them is making the request separately

Reply:
I think they shouldn't when user inputs branch, for example, we shouldn't show loading status on all fields, it should be just a branch. So I combined auth stuff in one debounce, and each other field should have its own one

Reply:
There are two problems with that:

* API response gives results for all fields at once
* The fields are strictly connected - if you will change the repository URI, it means that the branch may become invalid

Additionally, you don't have to show a loader for all of them, but still you might aggregate them in a single call

Reply:
Honestly don't see that much of a problem. Yes, fields are connected, but still branch validation will be correctly displayed if URI is invalid. I will try to see if I can resolve loaders. Not sure right now how to identify which exact field was changed and display loader for it.

Reply:
If it will be hard - ignore it; this part, as long as the race condition will be solved, should mostly work okay anyway (except unnecessary calls and sometimes some flaky interface)